### PR TITLE
[IOPID-636][IOPID-637] : lock/unlock profile handlers 

### PR DESCRIPTION
**Pull request by pelliccm**
Implementation of the lock and unlock session handlers.

#### List of Changes

- Add lock-session handler
- Add unlock-session handler
- Unit tests of both

#### Motivation and Context

Change required to let user perform the actions of locking and/or unlocking their profile.

#### How Has This Been Tested?

- Unit tests
- Postman in dev environment

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by pelliccm** [comment]
Part of this changes come from [pr of task 615](https://github.com/pagopa/io-web-profile-backend/pull/8)

File: openapi/io-functions-fast-login.yaml
Comment (arcogabbo) [inline openapi/io-functions-fast-login.yaml]: as far as i remember this endpoint can also return a 429 which i don't see below in the responses. cc. @BurnedMarshal

File: openapi/io-functions-fast-login.yaml
Comment (arcogabbo) [inline openapi/io-functions-fast-login.yaml]: as stated in the confluence document, this endpoint should be `session-state` instead of `sessions-state`
```suggestion
  /v1/session-state:
```

File: Ping/handler.ts
Comment (arcogabbo) [inline Ping/handler.ts]:
```suggestion
```

File: openapi/io-functions-fast-login.yaml
Comment (arcogabbo) [inline openapi/io-functions-fast-login.yaml]: I think all endpoints should be prefixed with `/api` in the spec

File: openapi/io-functions-fast-login.yaml
Comment (arcogabbo) [inline openapi/io-functions-fast-login.yaml:117]: the body has to be required
```suggestion
      requestBody:
        required: true
```

File: openapi/io-functions-fast-login.yaml
Comment (arcogabbo) [inline openapi/io-functions-fast-login.yaml:152]:
```suggestion
      requestBody:
        required: true
```

File: openapi/io-functions-fast-login.yaml
Comment (arcogabbo) [inline openapi/io-functions-fast-login.yaml:189]:
```suggestion
      requestBody:
        required: true
```

File: openapi/io-functions-fast-login.yaml
Comment (arcogabbo) [inline openapi/io-functions-fast-login.yaml:228]:
```suggestion
      requestBody:
        required: true
```

File: LockSession/handler.ts
Comment (arcogabbo) [inline LockSession/handler.ts]:
```suggestion
```

File: LockSession/handler.ts
Comment (arcogabbo) [inline LockSession/handler.ts]: I suggest to remove this lines and add a TE.bind in the pipeline below
```suggestion
```

File: LockSession/handler.ts
Comment (arcogabbo) [inline LockSession/handler.ts]:
```suggestion
export const lockSessionHandler = (
  client: LockSessionClient
): ILockSessionHandler => (
  reqJwtData,
  reqPayload
): ReturnType<ILockSessionHandler> =>
  pipe(
    TE.Do,
    TE.bind("user_data", () => TE.of(reqJwtData)),
    TE.bind("unlock_code", () => TE.of(reqPayload.unlock_code)),
    TE.chain(
      TE.fromPredicate(
        ({ user_data }) => canLock(user_data),
        ({ user_data }) =>
          getResponseErrorForbiddenNotAuthorized(
            `Could not perform lock-session. Required SpidLevel at least: {${SpidLevel.L2}}; User SpidLevel: {${user_data.spid_level}}`
          )
      )
    ),
    TE.chainW(({ user_data, unlock_code }) =>
      TE.tryCatch(
        () =>
          client.lockUserSession({
            body: {
              fiscal_code: user_data.fiscal_number,
              unlock_code
            }
          }),
```

File: openapi/io-functions-fast-login.yaml
Comment (arcogabbo) [inline openapi/io-functions-fast-login.yaml:1]: in this file there are some errors in refs. some of them starts with `#components` instead of `#/components`.

File: UnlockSession/handler.ts
Comment (arcogabbo) [inline UnlockSession/handler.ts]:
```suggestion
```

File: UnlockSession/handler.ts
Comment (arcogabbo) [inline UnlockSession/handler.ts]: I believe a similar logic with TE.bind can be applied here. see comments in the handler inside `LockSession` folder

File: UnlockSession/__test__/handler.test.ts
Comment (arcogabbo) [inline UnlockSession/__test__/handler.test.ts:53]:
```suggestion
    expect(unlockSessionMock).toHaveBeenCalledTimes(1);
    expect(unlockSessionMock).toHaveBeenCalledWith({
      body: {
        fiscal_code: aValidL2User.fiscal_number,
        unlock_code: aValidL2Payload.unlock_code
      }
    });
    expect(res).toMatchObject({
      kind: "IResponseSuccessNoContent"
    });
```

File: UnlockSession/__test__/handler.test.ts
Comment (arcogabbo) [inline UnlockSession/__test__/handler.test.ts:72]:
```suggestion
    expect(unlockSessionMock).toHaveBeenCalledTimes(1);
    expect(unlockSessionMock).toHaveBeenCalledWith({
      body: {
        fiscal_code: aValidL2User.fiscal_number
      }
    });
```

File: UnlockSession/__test__/handler.test.ts
Comment (arcogabbo) [inline UnlockSession/__test__/handler.test.ts:42]:
```suggestion
describe("UnlockSession", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
```

File: LockSession/__test__/handler.test.ts
Comment (arcogabbo) [inline LockSession/__test__/handler.test.ts:1]: similar additions that I suggested for the handler tests for `UnlockSession` can be applied here.

**Review by arcogabbo — CHANGES_REQUESTED** [review]
nice work. I strongly believe we can enhance the handler tests with error scenarios before merge. cc. @BurnedMarshal

File: UnlockSession/handler.ts
Comment (arcogabbo) [inline UnlockSession/handler.ts]:
```suggestion
```

File: UnlockSession/handler.ts
Comment (arcogabbo) [inline UnlockSession/handler.ts]: here we are injecting an error to the right part of the taskeither using TE.map . I suggest taking a slightly different approach with TE.chain as we did in the `introspectionCall` function in another file

File: LockSession/handler.ts
Comment (arcogabbo) [inline LockSession/handler.ts]: in this line we are injecting an error to the right part of the taskeither. as stated in another comment, I suggest a  different approach

**Review by arcogabbo — CHANGES_REQUESTED** [review]

File: UnlockSession/handler.ts
Comment (arcogabbo) [inline UnlockSession/handler.ts]:
```suggestion
```

File: utils/enums/SpidLevels.ts
Comment (BurnedMarshal) [inline utils/enums/SpidLevels.ts]: `ReadOnlyNonEmptySet` should be a more accurate type but only for the name.

```suggestion
import { ReadOnlyNonEmptySet } from "@pagopa/ts-commons/lib/types";

export enum SpidLevel {
  L1 = "https://www.spid.gov.it/SpidL1",
  L2 = "https://www.spid.gov.it/SpidL2",
  L3 = "https://www.spid.gov.it/SpidL3"
}

type SpidComparison = Record<SpidLevel, ReadOnlyNonEmptySet<SpidLevel>>;
```

**Review by arcogabbo — COMMENTED** [review]
it's okay IMO. as i stated in the first review,
> I strongly believe we can enhance the handler tests with error scenarios before merge

let's also wait for @BurnedMarshal approval

File: LockSession/handler.ts
Comment (BurnedMarshal) [inline LockSession/handler.ts]: Accordingly with the functional analysis the `409` conflict error response should be forwarded to the client. If the client retrieve this response only in a retry tentative it will be considered as a success, otherwise will be show an error.
```suggestion
          switch (response.status) {
            case 204:
              return TE.right(ResponseSuccessNoContent());
            case 409:
              return TE.left<
                ILookSessionErrorResponses,
                IResponseSuccessNoContent
              >(ResponseErrorConflict("Session was already locked"));
            default:
              return TE.left<
                ILookSessionErrorResponses,
                IResponseSuccessNoContent
              >(
                ResponseErrorInternal(
                  `Something gone wrong. Response Status: {${response.status}}`
                )
              );
          }
```
defining
```typescript
type ILookSessionErrorResponses =
  | IResponseErrorConflict
  | IResponseErrorForbiddenNotAuthorized
  | IResponseErrorInternal;

type ILockSessionHandler = (
  user: IHslJwtPayloadExtended,
  payload: LockSessionData
) => Promise<IResponseSuccessNoContent | ILookSessionErrorResponses>;
```

File: LockSession/handler.ts
Comment (BurnedMarshal) [inline LockSession/handler.ts]: We must add the context to be able to log errors.
```suggestion
  const middlewaresWrap = withRequestMiddlewares(
    ContextMiddleware(),
    verifyUserEligibilityMiddleware(config),
    hslJwtValidationMiddleware(config),
    RequiredBodyPayloadMiddleware(LockSessionData)
  );
```

File: UnlockSession/handler.ts
Comment (BurnedMarshal) [inline UnlockSession/handler.ts]: This is an optional value, we can use the `Option` object to map this scenario, avoiding to handle undefined values.

File: UnlockSession/handler.ts
Comment (BurnedMarshal) [inline UnlockSession/handler.ts]: The unlock should never returns conflicts.
Instead we must handle the `403` forbidden error.

File: openapi/io-functions-fast-login.yaml
Comment (BurnedMarshal) [inline openapi/io-functions-fast-login.yaml]: This scenario should be removed

File: package.json
Comment (BurnedMarshal) [inline package.json]: This generate also the client, rename removing the `definitions` prefix

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]
Can be useful add some error logs especially for promise rejection when we use the client to calling a service.
[Here](https://github.com/pagopa/io-functions-lollipop/blob/master/GetAssertion/handler.ts#L94) an example using the new `@pagopa/winston-ts`.
You can configure the logger using `useWinstonFor` with the default logger id, [here](https://github.com/pagopa/io-functions-lollipop/blob/b5f83f5a46f77f29c74ff0a813e2585f1c80f0be/GetAssertion/index.ts#L52) an example.

**Review by BurnedMarshal — APPROVED** [review]
LGTM

**Review by arcogabbo — APPROVED** [review]

**Pull request merged**
